### PR TITLE
Display a static chess state on AR

### PR DESCRIPTION
**Pull request by KurohanaJuri**
This PR change the AR display, we can now display a board with pieces with colour.

Big changes : 
- The models are loaded and added into the view at creation
    - This fix an old bug that makes impossible to use the function `smooth` to move an object from a point A to B
  
This PR also centralize the graphic content into one place. 

**Next step :** 

For the moment, the scene only display a static board. We need some modification to be able to update the AR game with the incoming new state. We need to add a listener that will change one piece position to the new location. To do that, we have 2 possibility : 
- "Spawning" : We change the position without animation, the piece disappear to spawn on the new position
- With animation : We can do a smooth moving animation. This solution is a bit more challenging if we don't want a "cheap" effect (Piece that collide with other). The best animation is a bell-shaped trajectory, which is tricky.

--------------------------
**Possible Enhancement :** 
- Allow user to customize the piece colour
    - The colour is set dynamically, we can imagine having multiple preset that the user can choose (or pay to have it ;D)
    - The board customization is a bit trickier, it's not a uniform colour, hence the "colour" is a texture come with the models, we can have multiple board skin in different files
- We can also have another set of pieces
- Display match information in AR (player turn, end of the game,....)

### CODE REVIEW

File: mobile/src/main/java/ch/epfl/sdp/mobile/ui/ar/ChessScene.kt [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+ * - The board as the root node
+ * - A number of chess pieces depending on the game state
+ *
+ * @param context The context used to load the 3d models
+ * @param lifecycleScope A scope that is used to launch the model loading
+ * @param board The board that contains the displayed game state
+ */
+class ChessScene(
+    private val context: Context,
+    private val lifecycleScope: LifecycleCoroutineScope,
```
Comment (alexandrepiveteau) [inline mobile/src/main/java/ch/epfl/sdp/mobile/ui/ar/ChessScene.kt:24]: Question : is a `LifecyleCoroutineScope` needed, or would a `CoroutineScope` be enough ?

**Review by alexandrepiveteau — COMMENTED** [review]
Looks good, and it's great that the coverage is above 80%.

I feel like there's a small violation of the MVVM pattern though - the view should not use a `Game` directly, since it's a Model object; instead, I think the approach should be similar to the one used in the 2d chessboard. I was thinking we may make a refactoring of the `ChessBoardState` in the next sprint so we can extract the common logic for loading / managing the game state between the 2d and 3d boards. What are your thoughts on this ?

File: mobile/src/main/java/ch/epfl/sdp/mobile/ui/Colors.kt [hunk 2]
```diff
@@ -24,4 +25,8 @@ object PawniesColors {
   val Green800 = Color(0xFF356859)
   val Orange500 = Color(0xFFFD5523)
   val Orange200 = Color(0xFFF8A68D)
+
+  // ArModel Color
+  val White = Vector3(1f, 0.99f, 0.94f)
+  val Black = Vector3(53 / 255f, 56 / 255f, 57 / 255f)
```
[the remark this answers is not in the document]
Comment (KurohanaJuri) [reply mobile/src/main/java/ch/epfl/sdp/mobile/ui/Colors.kt:31]: Yeah, I wasn't sure where to put these values.

File: mobile/src/main/java/ch/epfl/sdp/mobile/ui/ar/ChessScene.kt (hunk 1 above)
Comment (KurohanaJuri) [reply mobile/src/main/java/ch/epfl/sdp/mobile/ui/ar/ChessScene.kt:24]: The method used to load 3D models request a `LifecyleCoroutineScope`

File: mobile/src/main/java/ch/epfl/sdp/mobile/ui/Colors.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (KurohanaJuri) [reply mobile/src/main/java/ch/epfl/sdp/mobile/ui/Colors.kt:33]: Fixes in https://github.com/epfl-SDP/android/pull/209/commits/39de33c9f94caa53fbe313d6c8ad389b3dbd2d56

File: mobile/src/androidTest/java/ch/epfl/sdp/mobile/test/ui/ar/ChessSceneTest.kt [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+  fun given_emptyState_when_initChessScene_Then_rootIsNotNull() {
+    val board = Game.create()
+    val chessScene =
+        ChessScene(rule.activity.applicationContext, rule.activity.lifecycleScope, board.board)
+
+    assertNotEquals(null, chessScene.boardNode)
+  }
+
+  @Test
+  fun give_chessScene_when_scaled_then_rootHasCorrectScaleVector() {
```
Comment (matt989253) [inline mobile/src/androidTest/java/ch/epfl/sdp/mobile/test/ui/ar/ChessSceneTest.kt:27]:
```suggestion
  fun given_chessScene_when_scaled_then_rootHasCorrectScaleVector() {
```

**Review by matt989253 — CHANGES_REQUESTED** [review]

**Comment by matt989253** [comment]
A few things to change, but looks pretty good! It's awesome that you managed to bring the coverage above 80%!

File: mobile/src/main/java/ch/epfl/sdp/mobile/ui/ar/ArScreen.kt [hunk 4]
```diff
@@ -166,14 +47,19 @@ fun ArScreen(modifier: Modifier = Modifier) {
...
         fun anchorOrMoveBoard(anchor: Anchor) {
+
           // Add only one instance of the node
-          if (!arSceneView.children.contains(currentBoard)) {
-            arSceneView.addChild(currentBoard)
+          if (!arSceneView.children.contains(chessScene!!.boardNode)) {
+            arSceneView.addChild(chessScene!!.boardNode)
           }
-          currentBoard.anchor = anchor
+          chessScene!!.boardNode.anchor = anchor
```
[the remark this answers is not in the document]
Comment (KurohanaJuri) [reply mobile/src/main/java/ch/epfl/sdp/mobile/ui/ar/ArScreen.kt]: Fixed in [c252ce1](https://github.com/epfl-SDP/android/pull/209/commits/c252ce1b6eb1f645c76e36e17e5f1e5a8334d0a3)

File: mobile/src/main/java/ch/epfl/sdp/mobile/ui/ar/ArScreen.kt [hunk 5]
```diff
@@ -2,162 +2,43 @@ package ch.epfl.sdp.mobile.ui.ar
...
-fun ArScreen(modifier: Modifier = Modifier) {
-
-  var pawnNode by remember { mutableStateOf<ArModelNode?>(null) }
-  var boardNode by remember { mutableStateOf<ArModelNode?>(null) }
-
-  var arBoard by remember { mutableStateOf<ArBoard?>(null) }
-
-  // FIX ME : Prototype purpose only, need to be replace with the [ChessBoardState]
-  val piecePosition = remember { ChessBoardState.Position(1, 1) }
+fun ArScreen(game: Game, modifier: Modifier = Modifier) {
```
[the remark this answers is not in the document]
Comment (KurohanaJuri) [reply mobile/src/main/java/ch/epfl/sdp/mobile/ui/ar/ArScreen.kt:29]: Fixed in [c252ce1](https://github.com/epfl-SDP/android/pull/209/commits/c252ce1b6eb1f645c76e36e17e5f1e5a8334d0a3)

**Review by matt989253 — APPROVED** [review]
LGTM!

**Review by alexandrepiveteau — APPROVED** [review]

**Pull request merged**